### PR TITLE
Refactored allocator module

### PR DESCRIPTION
**Pull request by oierlauzi**
This refactor is a result of recent changes in the core interfaces:

- Improved using buffers across multiple queues (record_queue re-implementation in a centralized manner)
- Buffers do not have an assotiated type: They are allocated based on size and alignment requirements.
- Using this opportunity to separate functions of various classes in the allocator library

### CODE REVIEW

**Comment by oierlauzi** [comment]
Sonarcloud warns about code duplication. Skipping for now.

File: include/xmipp4/cuda/compute/allocator/cuda_deferred_memory_block_release.inl [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        else
+        {
+            events.splice_after(
+                events.cbefore_begin(),
+                m_event_pool, 
+                m_event_pool.cbefore_begin()
+            );
+        }
+
+        events.front().signal(*queue);
```
Comment (MartinSalinas98) [inline include/xmipp4/cuda/compute/allocator/cuda_deferred_memory_block_release.inl:90]: Inconsistent use of record & signal?

File: include/xmipp4/cuda/compute/allocator/cuda_memory_block_usage_tracker.hpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    /**
+     * @brief Add a queue to the inventory.
+     * 
+     * The queue is only added if it is different to owner of the
+     * provided block and it has not been added previously.
+     * 
+     * @param block The cuda memory block.
+     * @param queue Queue where the block has been used.
+     */
+    void record_queue(const cuda_memory_block &block, cuda_device_queue &queue);
```
Comment (MartinSalinas98) [inline include/xmipp4/cuda/compute/allocator/cuda_memory_block_usage_tracker.hpp:70]: maybe `add_queue` is a bit more descriptive

File: tests/unitary/src/compute/allocator/test_cuda_memory_block.cpp [hunk 3]
```diff
@@ -38,9 +38,11 @@ TEST_CASE( "construct cuda_memory_block", "[cuda_memory_block]" )
 {
     const std::uintptr_t ptr_value = 0xDEADBEEF;
     auto *const ptr = reinterpret_cast<void*>(ptr_value);
+    const std::uintptr_t queue_value = 0xA7EBADF0D;
```
Comment (MartinSalinas98) [inline tests/unitary/src/compute/allocator/test_cuda_memory_block.cpp:41]: Where does this value come from?

**Review by MartinSalinas98 — CHANGES_REQUESTED** [review]

File: tests/unitary/src/compute/allocator/test_cuda_memory_block.cpp (hunk 3 above)
Comment (oierlauzi) [reply tests/unitary/src/compute/allocator/test_cuda_memory_block.cpp:41]: Arbitrary. I usually pick hexpeak values for pointer addresses.

Comment (MartinSalinas98) [reply tests/unitary/src/compute/allocator/test_cuda_memory_block.cpp:41]: But how does that work? I mean, are the hex addresses deterministic? Do you always get the same address regardless of the machine & execution?

Comment (oierlauzi) [reply tests/unitary/src/compute/allocator/test_cuda_memory_block.cpp:41]: The address is a constant. I pick readable values (0xDEADBEEF=dead beef 0xA7EBADF0D=ate bad food) for easier traceability in debugging. Accessing these addresses is generally illegal, but here (and other places) they are not accessed, they are used as unique identifiers.

**Review by MartinSalinas98 — APPROVED** [review]

**Pull request merged**